driver-core/src/main/java/com/datastax/driver/core/Frame.java

```diff
@@ -201,19 +201,32 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf buffer, List<Object> ou
             // version first bit is the "direction" of the frame (request or response)
             version = version & 0x7F;
 
-            DecoderForStreamIdSize decoder = (version >= 3) ? decoderV3 : decoderV1;
+            DecoderForStreamIdSize decoder = decoderForStreamIdSize.get();
+            if (decoder == null) {
```

Comment:
That will work in the general case, the main reason for restricting to one `DecoderForStreamStreamSize` is that if you reach the frame length limit, the `DecoderForStreamIdSize` tracks the remaining bytes to consume and discards them until remnant of the frame is discarded.  Although in general you should only expect frames sharing the same protocol version, and if that didn't happen the code I have here would fail in unexpected ways anyways, so maybe the best solution is to do as you suggest.